### PR TITLE
Normative: Add Array/%TypedArray%.prototype.includes

### PR DESCRIPTION
**Pull request by domenic**
Includes editorial fixes to inconsistencies between nearby algorithms or algorithms with similar constructs. Happy to un-do any of those but they seem like improvements to me.

### CODE REVIEW

**Comment by ljharb** [comment]
(ﾉ◕ヮ◕)ﾉ*:・ﾟ✧ woot

File: spec.html [hunk 1]
```diff
@@ -29748,6 +29748,45 @@
...
+        <emu-alg>
+          1. Let _O_ be ? ToObject(*this* value).
+          1. Let _len_ be ? ToLength(? Get(_O_, `"length"`)).
+          1. If _len_ is 0, return *false*.
+          1. Let _n_ be ? ToInteger(_fromIndex_). (If _fromIndex_ is *undefined*, this step produces the value 0.)
+          1. If _n_ ≥ 0, then
+            1. Let _k_ be _n_.
+          1. Else _n_ < 0,
+            1. Let _k_ be _len_ + _n_.
+            1. If _k_ < 0, then let _k_ be 0.
```
Comment (anba) [inline spec.html]: Nit: Remove "then".

File: spec.html [hunk 2]
```diff
@@ -29760,14 +29799,14 @@
...
           1. If _len_ is 0, return -1.
-          1. If argument _fromIndex_ was passed, let _n_ be ? ToInteger(_fromIndex_); else let _n_ be 0.
+          1. Let _n_ be ? ToInteger(_fromIndex_). (If _fromIndex_ is *undefined*, this step produces the value 0.)
           1. If _n_ &ge; _len_, return -1.
           1. If _n_ &ge; 0, then
             1. Let _k_ be _n_.
-          1. Else _n_&lt;0,
-            1. Let _k_ be _len_ - abs(_n_).
+          1. Else _n_ &lt; 0,
+            1. Let _k_ be _len_ + _n_.
```
Comment (anba) [inline spec.html]: Also remove abs() in Array.prototype.lastIndexOf()?

**Comment by domenic** [comment]
Fixed @anba's review comments and rebased

**Comment by allenwb** [comment]
regarding the following change to indexOf:

> 
-   -    1. If argument _fromIndex_ was passed, let _n_ be ? ToInteger(_fromIndex_); else let _n_ be 0.
+  +   1. Let _n_ be ? ToInteger(_fromIndex_). (If _fromIndex_ is *undefined*, this step produces the value 0.)

I rather don't like it, because the corresponding line in lastIndexOf actually requires the `if`. By making this change we would loose a bit of the parallel construction between  indexOf and lastIndexOf.  This will make it harder for anyone who is doing a line by line comparison of the two algorithms to the essential differences.

A more general point:  Once somebody (like most of us) become very familiar with the spec. things like the fact that ToInteger(undefined) is 0 becomes ingrained. But people (like many implementors) who only occasionally reference the spec. do not have that level of familiarity. Such readers are better served , in cases like this, if we simply state the result rather than expressing it in terms of an abstract operation. 

A similar readability issue applies to the suggested change of `len - abs(n)` to `len + *`.  The former can be understood without looking at surrounding context.  The latter requires understand that this expression is guarded by `n <  0` predicated.  The abs from also emphasizes the fact that we are decreasing the length.

**Comment by domenic** [comment]
> I rather don't like it, because the corresponding line in lastIndexOf actually requires the if. By making this change we would loose a bit of the parallel construction between indexOf and lastIndexOf. This will make it harder for anyone who is doing a line by line comparison of the two algorithms to the essential differences.

Right, but it brings it more in line with the rest of the spec. And it makes the essential difference clearer: lastIndexOf relies on (in self-hosted terms) `arguments.length`, whereas indexOf does not.

> A similar readability issue applies to the suggested change of len - abs(n) to len + *

We've had this discussion before, and in fact implementers were the ones that suggested the change. They preferred eliminating the excessive abs() call, making the algorithm more straightforward and reflective of the math, even if it requires them to look up a line to understand whether the result will be greater or less than len.

**Comment by anba** [comment]
Replacing abs() with addition also aligns indexOf and lastIndexOf with the other methods (copyWithin, fill, slice and splice).

**Comment by bterlson** [comment]
I think this is good. Allen raises valid points but in this case I don't think readability suffers terribly. Working on landing this now.

**Comment by bterlson** [comment]
Committed as c8102ed. Thanks! :sparkle:

**Pull request closed without merging**